### PR TITLE
Add Pathlib

### PR DESCRIPTION
**Pull request by palaviv**
This change add `pathlib` and changes many `os` method to support path like objects.
Many of the tests fail because Rust does not support `Cstring` with `nul`. Not sure how to bypass that. Would appreciate suggestion on how to solve this.

### CODE REVIEW

**Comment by coolreader18** [comment]
I did implement TryFromObject for CString recently, but you could also just have a nulcheck function that makes sure a string doesn't have a nul char in it, and returns a ValueError if it does, then just call `nulcheck(s, vm)?;`. (Or maybe a vm method?)

**Comment by palaviv** [comment]
> I did implement TryFromObject for CString recently, but you could also just have a nulcheck function that makes sure a string doesn't have a nul char in it, and returns a ValueError if it does, then just call `nulcheck(s, vm)?;`. (Or maybe a vm method?)

The problem is that CPython support `nul` char while rust throws exception.
```py
os.stat("/dev/null\x00")
```
Should not throw an error.

**Comment by coolreader18** [comment]
Ah, I see. Hmm, that is tricky.

File: vm/src/stdlib/os.rs [hunk 1]
```diff
@@ -1087,21 +1109,16 @@ fn os_chmod(
...
-        || objtype::issubclass(&path.class(), &vm.ctx.bytes_type())
-    {
-        Ok(path)
-    } else {
-        Err(vm.new_type_error(format!(
-            "expected str or bytes object, not {}",
-            path.class()
-        )))
-    }
+fn os_fspath(path: PyPathLike, vm: &VirtualMachine) -> PyResult {
```
Comment (youknowone) [inline vm/src/stdlib/os.rs:1112]: is there `Err` path in this function?

**Comment by palaviv** [comment]
> It seems only trailing `\x00` exists in tests. Trimming them before making `PyPathLike` would work?

This will actually give the opposite result. As `/dev/null` should be different then `/dev/null\x00`

 > This is not specifically related to this PR, but while reading, i found i have no idea why PyPathLike is not PyPath but Like

Thats how they are called in CPython.

**Comment by youknowone** [comment]
I try to test this problem a bit, but I can't make null-included path working. Could you describe your working environment?

macOS, python 3.8.1, APFS
```
>>> open('abc\x00', 'w')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: embedded null byte
>>> import os
>>> os.stat('/dev/null\x00')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: embedded null byte
>>>
```

Linux,  Python 3.5.2, ext4
```
>>> open('abc\x00', 'w')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: embedded null byte
>>> import os
>>> os.stat('/dev/null\x00')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: embedded null byte
```
```

**Comment by palaviv** [comment]
If you will look at the tests. For example https://github.com/RustPython/RustPython/pull/1840/files#diff-6bc3eb7fdef7557c945c66fb0b60a134R1921

**Comment by palaviv** [comment]
@coolreader18 @youknowone can this be pushed and we will solve the `nul` later?

**Comment by youknowone** [comment]
It seems `Vec<u8>` is the reasonble internal data of path types. And they may be used to create ffistring in future when they are actually used. Is it possible to fix now? or do you want to merge this for right now as it is?

**Comment by palaviv** [comment]
> It seems `Vec<u8>` is the reasonble internal data of path types. And they may be used to create ffistring in future when they are actually used. Is it possible to fix now? or do you want to merge this for right now as it is?

I am not sure I understand... Would you like to change the `OsString` to `Vec<u8>`?

**Comment by youknowone** [comment]
I guess so. I didn't check cpython implementation well yet, but it seems it is more close to `Vec<u8>` than `OsString`. Or `enum { Bytes(Vec<u8>), Str(String) }` if there are difference by str/bytes.

**Comment by coolreader18** [comment]
@palaviv I think fixing it later would be fine, it's not too critical to pathlib, and I think it would rarely happen in practice.

**Comment by palaviv** [comment]
I have merged this and we can do the proposed changes in the future

**Comment by youknowone** [comment]
I checked about PyPathLike impl but it revealed enum is not fit for the use case.

**Pull request merged**